File: Backend/apps/attendance/permissions.py
from hamamooz.apps.accounts.access import user_has_role
from hamamooz.apps.accounts.models import Role

ATTENDANCE_WRITERS = [
    Role.SYSTEM_ADMIN,
    Role.ORGANIZATION_ADMIN,
    Role.SCHOOL_MANAGER,
    Role.EDUCATIONAL_DEPUTY,
    Role.OPERATOR,
    Role.TEACHER,
]

ATTENDANCE_REVIEWERS = [
    Role.SYSTEM_ADMIN,
    Role.ORGANIZATION_ADMIN,
    Role.SCHOOL_MANAGER,
    Role.EDUCATIONAL_DEPUTY,
]

ATTENDANCE_POLICY_MANAGERS = ATTENDANCE_REVIEWERS

BROAD_ATTENDANCE_ROLES = [
    Role.SYSTEM_ADMIN,
    Role.ORGANIZATION_ADMIN,
    Role.SCHOOL_MANAGER,
    Role.EDUCATIONAL_DEPUTY,
    Role.OPERATOR,
]


def can_manage_session(user, session):
    if user_has_role(
        user,
        BROAD_ATTENDANCE_ROLES,
        organization_id=session.organization_id,
        school_id=session.school_id,
    ):
        return True
    return bool(
        session.scope == session.Scope.PERIOD
        and session.course_offering_id
        and session.course_offering.teacher_id == user.id
        and user_has_role(
            user,
            [Role.TEACHER],
            organization_id=session.organization_id,
            school_id=session.school_id,
        )
    )

File: Frontend/src/app/permissions.ts
import { store } from './store.js';
import type { Role } from '../api/types.js';

export const administrativeRoles: Role[] = ['system_admin', 'organization_admin', 'school_manager'];
export const managementReadRoles: Role[] = [...administrativeRoles, 'educational_deputy', 'student_affairs_deputy'];
export const organizationManagementRoles: Role[] = ['system_admin', 'organization_admin'];
export const policyManagementRoles: Role[] = ['system_admin', 'organization_admin', 'school_manager', 'educational_deputy', 'student_affairs_deputy'];
export const broadEducationRoles: Role[] = ['system_admin', 'organization_admin', 'school_manager', 'educational_deputy', 'student_affairs_deputy', 'operator'];
export const counselingRoles: Role[] = ['counselor'];
export const guidanceRoles: Role[] = [...broadEducationRoles, 'guide_teacher'];
export const teacherWriteRoles: Role[] = [...broadEducationRoles, 'teacher'];
export const curriculumManagementRoles: Role[] = ['system_admin', 'organization_admin', 'educational_deputy'];

export function activeRoles(): Role[] {
  return [...new Set((store.state.user?.role_assignments ?? []).filter(item => item.is_active).map(item => item.role))];
}

export function hasAnyRole(roles?: Role[]): boolean {
  if (!roles?.length) return true;
  const current = activeRoles();
  return current.includes('system_admin') || roles.some(role => current.includes(role));
}

/**
 * Check an explicit role boundary without the system-admin convenience
 * override. This is used for confidential domains whose backend permission
 * model intentionally does not grant global administration access.
 */
export function hasExactRole(roles?: Role[]): boolean {
  if (!roles?.length) return true;
  const current = activeRoles();
  return roles.some(role => current.includes(role));
}

export function hasWriteScope(): boolean {
  return isSystemAdmin() || Boolean(store.state.scope.schoolId || store.state.scope.organizationId);
}

export function isSystemAdmin(): boolean {
  return activeRoles().includes('system_admin');
}

export function roleLabel(role: Role): string {
  const labels: Record<Role, string> = {
    system_admin: 'مدیر کل سامانه',
    organization_admin: 'مدیر مجموعه',
    school_manager: 'مدیر مدرسه',
    educational_deputy: 'معاون آموزشی',
    operator: 'اپراتور',
    teacher: 'دبیر',
    student_affairs_deputy: 'معاون امور دانش‌آموزی',
    counselor: 'مشاور',
    guide_teacher: 'معلم راهنما',
  };
  return labels[role];
}
